### PR TITLE
[14.0][ADD] l10n_th_sequence

### PR DESCRIPTION
**Pull request by sansirit**
As inspired by PR [244](https://github.com/OCA/l10n-thailand/pull/244), I create this module to implement another way.

This module refactor the `ir.sequence` allowing more legends. Then, it add the following legends:

* Current Buddhist Era (BE) Year with Century: `%(year_be)s`
* Current Buddhist Era (BE) Year without Century: `%(y_be)s`
* Quarter of the Year: `%(qoy)s`
* Company branch number (5-digit): `%(b5)s`
* Company branch number (4-digit): `%(b4)s`
* Company branch number (3-digit): `%(b3)s`
* Company branch number (2-digit): `%(b2)s`
* Company branch number (1-digit): `%(b1)s`

This module also add a `preview` field in the form view of the `ir.sequence`. So, a user can see and check the sequence configuration easily.

**Note**: This module does not depends on the `l10n_th_partner`, although it refer to the field `branch` of `res.company`.
if the module `l10n_th_partner` is not installed (`env.company` has no `branch` field), the company branch number will be computed as "00000".

This module works with `date` and `date_range`, but not the `range_end` from server_ux [PR](https://github.com/OCA/server-ux/pull/361).

### CODE REVIEW

**Comment by kittiu** [comment]
@sansirit thanks for the proposal. branch and qoy are good additions and want to push it forward,

1. By OCA way, we need to avoid using `override`, to respect the calling chain, the better way is to use a monkey patch addon. (the best is to propose change to Odoo S.A. but that is not easy to success). 
    - https://odoo-community.org/groups/contributors-15/contributors-129853?mode=date&date_begin=&date_end=, I also sent you an email as the link seem no good now.
2. We need it to work with `squence_range_end`, which is much needed for Thailand government context. I.e., 1 Oct 2564 is considered FY 2565 in Thai government, so range_end is needed.

Your proposed features set are not just for Thailand and should be in https://github.com/OCA/server-ux, so the rest of community can appreciate it (and to get more reviews/opinions there).

So, in my opinion to follow the OCA ways (small workable addon). I think the best set of modules are,

* **sequence_get_prefix_suffix_hook**
* sequence_range_end (already exist but need to follow the hook)
* sequence_buddhisht_era
* sequence_quarter_of_year
* sequence_company_branch

But if that is too much, we can propose like this,

* **sequence_get_prefix_suffix_hook**
* sequence_prefix_suffix_options (we can include all above options here).

WDYT?

cc @pedrobaeza you have any opinion about this?

**Note:**
* You can check this, for some more info about the first attempt: https://github.com/OCA/server-ux/pull/360
* Before I did `l10n_th_sequence`, I was trying to propose it as `server-ux/sequence_buddhist_year`, and want be able to use as `date_range_end_buddhist_year`, but I could find the way (we don't have the hook module yet). We need also `sequence_range_end` to use for Thai government, so I decide to move to l10n-thailand the way it is (I don't like but nonetheless).

**Comment by pedrobaeza** [comment]
Although you need to replicate some logic from upstream having an inner method, you can copy/paste such part in your method with your prefixes, call super, and the reapply your part, so you respect inheritance.

**Comment by kittiu** [comment]
> Although you need to replicate some logic from upstream having an inner method, you can copy/paste such part in your method with your prefixes, call super, and the reapply your part, so you respect inheritance.

Thanks! @pedrobaeza 
I think what Pedro say, it similar to what I did for the sequence_range_end -> https://github.com/ecosoft-odoo/server-ux/blob/14.0-add-sequence_range_end/sequence_range_end/models/ir_sequence.py#L19

**Comment by pedrobaeza** [comment]
Yes, but I think you should call always to super, as there can be more than one placeholder.

**Comment by sansirit** [comment]
> So, in my opinion to follow the OCA ways (small workable addon). I think the best set of modules are,
> 
> * **sequence_get_prefix_suffix_hook**
> * sequence_range_end (already exist but need to follow the hook)
> * sequence_buddhisht_era
> * sequence_quarter_of_year
> * sequence_company_branch
> 
> But if that is too much, we can propose like this,
> 
> * **sequence_get_prefix_suffix_hook**
> * sequence_prefix_suffix_options (we can include all above options here).
> 
> WDYT?

Creating a dedicated hook module is a good idea. I like that.
I am OK with either way. However, I prefer the later case that combined many options in a single module.

Regarding to the `sequence_range_end`, I have some questions about it. I see that the `range_end_` get information from the `date_to` field of model `ir.sequence.date_range`. Because each `ir.sequence.date_range` is tied to its `ir.sequence`. This architecture will require a system admin to create and maintain many `date_range` records. 

For the fiscal year information, Odoo has some hidden fields in the `res.company` model in the `account` module for that which are `fiscalyear_last_day` and `fiscalyear_last_month` (see [code here](https://github.com/odoo/odoo/blob/14.0/addons/account/models/company.py#L41)). Although, the `res.config` UI are hidden in the CE version, we OCA should have a module to make it usable. In addition, we can use `get_fiscal_year()` from 'odoo.tools.date_utils` to get information from those fields easily. By doing this, the system is more maintainable. 

WDYT?

**Comment by kittiu** [comment]
> I am OK with either way. However, I prefer the later case that combined many options in a single module.

I am ok too, for a single module. By the way, with that, and according to Pedro comment, I am not sure we still need the hook module.

> This architecture will require a system admin to create and maintain many date_range records.

I think we still need to use date_range, to conform with how Odoo use `date_from` when dealing with `range_year` prefix. Also, I think it is more granular (can be per document sequence) than using company.
Also, there is no worry about managing date_range manually every year, because it will be auto created (OCA also have module that fix standard Odoo issue).

In summary, my option, we still need to use `date_to` for `range_end`

**Comment by kittiu** [comment]
@sansirit do you plan to continue with this?

**Comment by sansirit** [comment]
Yes. I caught up in another project. I will post a new version in two weeks.

**Comment by kittiu** [comment]
> Yes. I caught up in another project. I will post a new version in two weeks.

Great, thanks! I was about to work on it. But it is always better to be done by the original author. Thanks again.

File: l10n_th_sequence/models/ir_sequence_date_range.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    def onchange_sequence_id(self):
+        self._compute_preview()
+
+    def _next(self):
+        return super(IrSequenceDateRangeExt, self.with_context(
+            ir_sequence_date_range_end=self.date_to
+        ))._next()
+
+    def _get_number_next_actual(self):
```
Comment (kittiu) [inline l10n_th_sequence/models/ir_sequence_date_range.py]: Please do not override method. I am not sure why this is needed.
Note: I only see slight diff in this line, `seq_id = "%03d_%03d" % (seq.sequence_id.ids[0], seq.ids[0])` not sure this is your intention.

File: l10n_th_sequence/models/ir_sequence.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        res.update(
+            {
+                "qoy": str((int(res["month"]) - 1) // 3 + 1),
+                "range_qoy": str((int(res["range_month"]) - 1) // 3 + 1),
+                "range_end_qoy": str((int(res["range_end_month"]) - 1) // 3 + 1),
+                "current_qoy": str((int(res["current_month"]) - 1) // 3 + 1),
+            }
+        )
+        branch = (
+            self.company_id.branch.zfill(5) if self.company_id.branch else ""
```
Comment (kittiu) [inline l10n_th_sequence/models/ir_sequence.py]:
```suggestion
            self.company_id.branch.zfill(5) if hasattr(self.company_id, "branch") else ""
```

Comment (kittiu) [reply l10n_th_sequence/models/ir_sequence.py]: Otherwise, we will have error if l10n_th_partner is not installed.

File: l10n_th_sequence/models/ir_sequence.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        res = {}
+        for key, fmt in sequences.items():
+            res[key] = effective_date.strftime(fmt)
+            res["range_" + key] = range_date.strftime(fmt)
+            res["range_end_" + key] = range_end_date.strftime(fmt)
+            res["current_" + key] = now.strftime(fmt)
+
+        return res
+
+    def _get_prefix_suffix(self):
```
Comment (kittiu) [inline l10n_th_sequence/models/ir_sequence.py]: Please don't override this method, we need to always call super(). You can do something like this?
https://github.com/OCA/server-ux/pull/361/files#diff-9e9a96594f4c35da625b41834dc4f4ed874f65560d36af78fe8fbb22e04317a4R14-R17

**Comment by kittiu** [comment]
I can see that your code already include `range_end_xxx` already. Can you also add description similar to this,
https://github.com/OCA/server-ux/pull/361/files#diff-f98996c2602eacbe89a2129921778a049ce2aba3cde33cb5ae8f499b6cb4a133R9

**Comment by kittiu** [comment]
Please correct test script and use pre-commit.

**Comment by kittiu** [comment]
Overall I test functionally, it works great!

**Comment by kittiu** [comment]
@sansirit 

I think this useful module is almost done. In fact I have the need to use it in project :)

1. Are you ok if I submit a PR to your branch, just to speed things up (if you really don't have time, otherwise let me know)
2. I think this module is beyond `l10n-thailand`. I rather propose it to `server-ux` WDYT?

**Comment by sansirit** [comment]
Yes I'm ok with you submit PR to my branch, it is your original idea after all.

I think the sequence preview part can put to server-ux.

**Comment by sansirit** [comment]
@kittiu please review this version. 

I've separated the modules into 6 modules.
- `l10n_th_sequence_refactored` is like a base module allowing the other modules to add more legends.
- `l10n_th_sequence_preview` adds the preview fields for both `ir.sequence` and `ir.sequence.date_range`
- `l10n_th_sequence_range_end` add the `range_end_` legends.
- `l10n_th_sequence_be`, `l10n_th_sequence_branch` and `l10n_th_sequence_qoy` add the BE year, company's branch and quarter of the year, respectively.

**Review by kittiu — APPROVED** [review]

**Comment by kittiu** [comment]
cc @ps-tubtim

**Comment by kittiu** [comment]
ping @ps-tubtim can you review this please.

**Comment by kittiu** [comment]
ping @ps-tubtim can you review this please.

**Comment by Saran440** [comment]
@sansirit Functional Test :+1: 

ผมมีเพิ่มเติมเรื่องการคำนวณ sequence ตอนใช้ date range ครับ

ปัญหาที่พบคือ ถ้าเลือก date from, date to คนละปีกัน ระบบจะคำนวณ sequence จาก date from เท่านั้น
ซึ่งเป็น standard odoo จึงเพิ่ม field range_date_selection
เพื่อให้ user สามารถเลือกได้ว่าต้องการให้ระบบออก sequence ของปีไหน

เช่น

Date From: 01/10/2021
Date To: 30/09/2022

ระบบจะออก sequence เป็นปี 2564 หรือ 64 เท่านั้น
แก้ไขให้ระบบสามารถออก sequence เป็นปี 2565 หรือ 65 แทนได้

โดยผมแก้ไปที่ PR -> https://github.com/sansirit/l10n-thailand/pull/1

Selection_001

**Review by Saran440 — APPROVED** [review]
